From 160a3d3ce4850f392af6737c71f7e4244b90460a Mon Sep 17 00:00:00 2001
From: Sarika Sharma <quic_sarishar@quicinc.com>
Date: Tue, 12 Mar 2024 08:23:44 +0530
Subject: [PATCH] wifi: ath12k: properly set bit for pdev mask for FW
 PPDU_STATS request

Currently request message type HTT_H2T_MSG_TYPE_PPDU_STATS_CFG
returns firmware PPDU_STATS, bit 8 to 15 is used as bitmask for
HTT_PPDU_STATS_CFG_PDEV_ID.
But bit 8 is reserved for SOC stats and pdev id starts from bit 9.
Due to this incorrect bit is set for that pdev and firmware will not
send PPDU_STATS corresponding to correct pdev therefore rate will not be
updated in host leading to below WARN_ON().

 Call trace:
[  913.580978]  airtime_link_metric_get+0x60/0xc4 [mac80211]
[  913.583150]  mesh_continuous_tx_fail_cnt+0x158/0x664 [mac80211]
[  913.588705]  mesh_rx_path_sel_frame+0x474/0x928 [mac80211]
[  913.594435]  ieee80211_mesh_rx_queued_mgmt+0x418/0x5d0 [mac80211]
[  913.599992]  ieee80211_ibss_leave+0x92c/0x1550 [mac80211]
[  913.606147]  process_one_work+0x194/0x270
[  913.611524]  worker_thread+0x200/0x314
[  913.615516]  kthread+0x140/0x150
[  913.619163]  ret_from_fork+0x10/0x18
[  913.622546] ---[ end trace b872654a493d5906 ]---

Hence change the bitmask for HTT_PPDU_STATS_CFG_PDEV_ID as bit 9
to 15 to properly fill the pdev id in request message.

Signed-off-by: Sarika Sharma <quic_sarishar@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp.h | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp.h b/drivers/net/wireless/ath/ath12k/dp.h
index e0679a6..72a1c78 100644
--- a/drivers/net/wireless/ath/ath12k/dp.h
+++ b/drivers/net/wireless/ath/ath12k/dp.h
@@ -801,7 +801,8 @@ struct htt_ppdu_stats_cfg_cmd {
 } __packed;
 
 #define HTT_PPDU_STATS_CFG_MSG_TYPE		GENMASK(7, 0)
-#define HTT_PPDU_STATS_CFG_PDEV_ID		GENMASK(15, 8)
+#define HTT_PPDU_STATS_CFG_SOC_STATS		BIT(8)
+#define HTT_PPDU_STATS_CFG_PDEV_ID		GENMASK(15, 9)
 #define HTT_PPDU_STATS_CFG_TLV_TYPE_BITMASK	GENMASK(31, 16)
 
 enum htt_ppdu_stats_tag_type {
-- 
2.34.1

